File: dsa/li05_binarysearch/java/L69_SqrtOfX.java

public class L69_SqrtOfX {

	public static void main(String[] args) {
		L69_SqrtOfX obj = new L69_SqrtOfX();
		int input = 8;
		System.out.println(obj.mySqrt(input));
	}

	// Problem: https://leetcode.com/problems/sqrtx/
	// Idea: Simple BS on range 1 to x
	// Time: O(logn)
	public int mySqrt(int x) {
		if(x == 0) return 0;

		int sqrt = -1;
		int low = 1, high = x;
		while(low <= high) {
			int mid = low + (high - low)/2;
			if(mid <= x/mid) {
				sqrt = mid;
				low = mid+1;
			}else {
				high = mid-1;
			}
		}
		return sqrt;
	}

}

File: dsa/concepts/05_binarysearch/python/l69_sqrt_of_x.py
"""
LeetCode 69 - Sqrt(x)

Idea:
Binary search on the range [1, x].
Use division instead of multiplication to avoid overflow.

Time Complexity: O(log n)
Space Complexity: O(1)
"""

def my_sqrt(x):
    if x == 0:
        return 0

    low, high = 1, x
    ans = 0

    while low <= high:
        mid = (low + high) // 2

        if mid <= x // mid:
            ans = mid          # mid*mid <= x
            low = mid + 1
        else:
            high = mid - 1

    return ans


# Driver code
if __name__ == "__main__":
    x = 8
    print(my_sqrt(x))  # 2
